### PR TITLE
New Algorithm

### PR DESCRIPTION
**Pull request by ds300**
Up until now, DerivableJS used a mark-and-sweep style algorithm to both propagate change and free unused memory.

This had a couple of minor performance implications, and one obscure-but-serious memory leak vulnerability (see #36).

This reimplementation of DerivableJS' core data structures removes the mark-and-sweep stuff, replacing it with a significantly simpler system inspired by Clojure's epochal STM.

The practical implications are: 
- atoms never ever store references to derivations. Only to reactors. This is ideal, and solves the memory management issue once and for all.
- the time complexity of updating atoms is now linear in the number of dependent reactors rather than in the number of dependent (derivations + reactors), and should be much more performant due to cache locality goodness.
- the cost of dereferencing derivables should be either similar or faster than before, but I'll have to do benchmarking to be sure of this.
- smaller file size (.min.gz has gone from **4,330 bytes** down to **3,718 bytes**. Makes all the difference.)

The progress of this reimplementation is as follows:
- [x] typescript prototype of epochal system to verify viability 
- [x] retrofit DJS with epochal system
- [x] get all the unit tests passing
- [x] trial epochal system in real-world projects already using DJS
- [x] modernize code base (npm build scripts, es6 modules + rollup, CI)
- [x] do white box testing. get 100% test coverage.
- [x] check performance and rework if necessary
- [ ] modify read-me to talk about push-pull as opposed to mark-and-sweep, like [Agera](https://github.com/google/agera)
- [ ] release

### CODE REVIEW

**Comment by liron00** [comment]
Sweet! Looking forward to replacing my workaround.

**Comment by ds300** [comment]
@liron00 I wonder if you'd be willing to test this on any of your projects? I managed to introduce a few sneaky bugs during this transition which only became apparent when testing on real projects.

I'm feeling confident that they're fixed now, and I've got 100% statement test coverage, but I'd still feel a lot better if your projects also work fine with the new stuff.

It's available on npm with the version tag `beta7`.

**Comment by liron00** [comment]
Sure I'll start using @beta7 and let you know how it goes.

**Comment by ds300** [comment]
Thanks :D

**Comment by ds300** [comment]
Quick update:

I did some preliminary benchmarking to find out how the new stuff compared to the old stuff, speed-wise. Turns out there was no improvement but, in fact, a slight worsening. I tried three or four approaches to bring it back up to speed, but none have been able to match DJS 0.12.1 - I've landed on using the fastest  of those approaches for now, which is not far away, but it would be nice to get that speed back at some point in the future.

The good news: DJS is still more than twice as fast as MobX! Although I'm not sure it would make much of a difference for ordinary workloads. So far I've been measuring the overhead of propagating changes to live reactors, which is normally dwarfed by the cost of computing new state and doing the side effects themselves.

Anyway I'm happy enough with the current state of things and will try to push forward with a release at the weekend.

**Comment by ds300** [comment]
New update,

I've largely re-written DJS yet again and now it is considerably faster than 0.12.x. I dumped the epochal verison tracking and replaced it with an obscenely simple system which sets up bi-directional relationships like before, but only for derivables which are touched by an active reactor. This means the bug in #36 is still fixed, but as a trade-off you don't get lazy caching goodness for derivables which are not used in reactions. I don't think anyone really dereferences derivables outside of reactions anyway, so not too much of a problem there.

I'm also introducing some breaking API changes for the 1.0.0-beta release coming up very soon. These are currently:

 - no more `.reactor` method.
 - no more whole-module `withEquality` function. but the `.withEquality` method remains 

Also some new stuff in the works

 - [ ] benchmarking against mobx (in progress)
 - [ ] `wrapPreviousState` function.
 - [ ] `.mReact` method
 - [ ] a quick-and-dirty practical gitbook guide to using the library.

I've also gotten rid of the big discourse about state management that I was writing in the README before. I gave a talk in London with more or less the same content but a slightly different focus involving React, which went extremely well so I'm going to turn that into a blog post on futurice.com and link to it from the readme.

File: package.json [hunk 1]
```diff
@@ -1,11 +1,24 @@
...
   "description": "Functional Reactive State for JavaScript & TypeScript",
   "author": "David Sheldrick",
   "files": [
     "dist"
   ],
+  "scripts": {
+    "build": "node scripts/build.js",
+    "test": "./node_modules/.bin/mocha --recursive",
+    "bench": "node scripts/bench.js",
+    "coverage": "./scripts/coverage.sh && ./node_modules/.bin/istanbul report --include=coverage-final.json text",
```
Comment (oskbor) [inline package.json]: ./node_modules/.bin is on the path when doing npm run.
so you can make it shorter like this:
"test": "mocha --recursive"

Comment (ds300) [inline package.json]: Oh, cool. Thanks for the heads up!

File: src/transactions.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+function TransactionContext(parent) {
+  this.parent = parent;
+  this.id2originalValue = {};
+  this.modifiedAtoms = [];
+}
+
+export function maybeTrack (atom) {
+  if (currentCtx !== null) {
```
Comment (oskbor) [inline src/transactions.js]: there is a function below called inTransaction() that you can use.
Hope u don't mind me commenting while i read through your code :)

Comment (ds300) [inline src/transactions.js]: Not at all! Who wouldn't want free code reviews? 😊

Comment (ds300) [inline src/transactions.js]: Actually I think I'm going to leave it as-is. That line gets called a *lot* and I'd like to save the compiler from having to inline it. Thanks anyhow :)

File: scripts/make-docs.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    fs.mkdirSync("docs");
+  }
+
+  var rootQ = null;
+  if (fs.existsSync(script)) {
+    console.log("docgen exists, skipping boot");
+    rootQ = {then: function (cb) { return cb(); }};
+  } else {
+    console.log("doicgen doesn't exist, compiling it...");
+    rootQ = execp("boot cljs", "./docgen");
```
Comment (oskbor) [inline scripts/make-docs.js]: this seems to require the cljs "boot" tool. Is it possible to avoid having to have the global "boot" installed?

**Comment by mruzekw** [comment]
I would love to see a CQRS example in the mix. Just to give an idea of how to structure a larger application. I notice there's a redux-like example here: https://github.com/andreypopp/react-derivable#fluxredux-like-unidirectional-data-flow

**Comment by ds300** [comment]
@mruzekw What kind of thing do you have in mind? I see Redux as a lightweight instance of the CQRS pattern. Although admittedly it doesn't do much on the query side of things until coupled with other libraries.

**Comment by mruzekw** [comment]
Does the flux challenge you've written use CQRS/redux?

**Comment by ds300** [comment]
Neither, although the approach is very similar indeed. The main difference from Redux is that the actions are implicit so they're not serializable. The reducer pattern is more-or-less intact. Porting the code to redux should be super easy.

**Comment by mruzekw** [comment]
I'll just give that a look and maybe convert it to Redux as an exercise.

**Comment by oskbor** [comment]
So close now! When will this land?

**Comment by ds300** [comment]
Not sure. I was planning to make nice new docs and a respectable-looking landing page before pushing out a 1.0.0 release. This was partly done, then summer happened and I got put on a very fulfilling project at work so my motivation to code on evenings and weekends has been waning.

However I've lately been learning the value of MVPs and thinking that it would be better to put out something more lean (docs-wise) to begin with at least. But still need to wait until I have some spare time at a weekend, which is looking like early October right now.

**Comment by zerkalica** [comment]
Can you shortly explain about breaking changes in feature release?  What sould be avoided in the current version?

**Comment by ds300** [comment]
The main breaking change is that the `Reactor` class and its associated API (`Derivation#reactor`) have gone away.

The top-level `withEquality` function was removed too.

Behaviour-wise, laziness is now only provided for 'active' derivations (which are upstream of an 'active' reactor). I might still change this back to the old behaviour (always lazy) before 1.0.0 is released if I can find a way to do it without impacting performance for active derivations.

I think everything else is the same.

**Comment by zerkalica** [comment]
I try to test 1.0.0-beta11, but catch strange bug: derive does not cache result, each get recalculate value again. It's some optimizations or bug?

```js
        const a = atom({p: 1})

        const b = a.derive((p) => {
            console.log('test')
        })

        b.get() // "test"
        b.get() // error: "test" again
```

**Comment by andreypopp** [comment]
> but I still would like to bring caching
back for all derivations somehow before it leaves beta.

Any idea how to achieve that? I'd be interested to work on a pull request.

**Comment by ds300** [comment]
@andreypopp The epochal system I started this PR with would work if kind-of superimposed over the current setup. That could end up being quite complex to implement and test, but it should add only a small amount of constant overhead. A PR would be super cool if you're feeling intrepid :)

**Comment by oskbor** [comment]
How is the rewrite coming along? ☺️

**Comment by ds300** [comment]
Oh gosh. Well the code is pretty tight (currently released as beta), but I
haven't had the extra-curricular moxie to finish off the docs and landing
page in recent months. I've got a big 5-week chunk of holiday coming up in
February; maybe that with replenish my OSS mana.

**Comment by oskbor** [comment]
Ill send you a mana potion for Christmas 🎅

**Comment by ds300** [comment]
This PR is ridiculous. I'm just gonna merge it into master and start doing feature and release branches.

**Pull request merged**